### PR TITLE
[BeamSearch]optimize key cache reordering

### PR DESCRIPTION
**Pull request by wangyems**
### Description
 

Replace onnxruntime::cuda::Transpose4DKernelParallelizeMultipleElementsPerThreadInInnermostDim() with custom transpose kernel in ReorderPastState(). The original implementation doesn't benefit from vectorized loading and coalesced accessing(write). and not fully utilize threads in the block. 

benchmarked with TNLGv4 model(batch=4, seq_len=4K)
transpose kernel speed up: ~1.9X (392 μs -> 206 μs)
overall reordering speedup: ~1.48X

Latency:
before:
image
after:
image

GPU matrix:
before:
image
after:
image

### CODE REVIEW

File: onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu [hunk 1]
```diff
@@ -1315,6 +1315,81 @@ template void BufferExpansionKernelLauncher(const int32_t* input,
...
+template <typename T>
+__global__ void ReorderPastStatesKernel(T* out_buffer,
+                                        const T* in_buffer,
+                                        int batch_size,
+                                        int num_heads,
+                                        int max_length,
+                                        int chunked_head_size,
+                                        int equv_chunk_size) {
+  //[B, N, max_length, H2(head_size/chunk_size), chunk_size] -> [B, N, H2(head_size/chunk_size), max_length, chunk_size]
+  __shared__ T tile[kSeqTileSize][kTileSize + 1];
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu:1330]: why `kTileSize + 1` instead of kTileSize?

Comment (wangyems) [reply onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu:1330]: it is to avoid bank conflicts.

**Comment by hariharans29** [comment]
These are nice gains. Are the gains noticeable with E2E generation when running batch size =4, seq length = 4k ?

File: onnxruntime/contrib_ops/cuda/transformers/generation_device_helper.cc [hunk 2]
```diff
@@ -79,27 +83,16 @@ Status ReorderPastState(
...
-                                                 past_state_shape[offset + 3] / chunk_size, past_state_shape[offset + 2],
-                                                 chunk_size};
-
-  // TODO(hasesh): Explore perf tuning for this Transpose operation
-  return onnxruntime::cuda::Transpose::DoTranspose(*static_cast<const cudaDeviceProp*>(cuda_device_prop), cuda_stream,
-                                                   cublas_handle, permutation,
-                                                   past_state_staging, past_state,
-                                                   &transpose_input_shape_override,
-                                                   &transpose_output_shape_override);
+  cuda::ReorderPastStatesKernelLauncher(past_state.MutableDataRaw(),
```
Comment (hariharans29) [inline onnxruntime/contrib_ops/cuda/transformers/generation_device_helper.cc:86]: Possibly a further enhancement (maybe for another PR) is to try and "fuse" this operation for all layers as opposed to doing staging memcpy + reorder layer by layer in a loop. For this, it needs some other changes - like allocating one flat buffer of past states for all layers (today this is not the case) + allocate one flat "staging" buffer for all layers (today we only allocate staging buffer for one layer and re-use for all layers) and then involking just 1 memcpy and 1 re-order kernel. This should save all the kernel launch overheads.

Comment (wangyems) [reply onnxruntime/contrib_ops/cuda/transformers/generation_device_helper.cc:86]: Actually this approach(https://github.com/microsoft/onnxruntime/tree/wangye/transpose_opt) was tried in the first place but saw performance drop. The kernel launch overheads can be ignored in the case of large input size.

**Comment by wangyems** [comment]
> These are nice gains. Are the gains noticeable with E2E generation when running batch size =4, seq length = 4k ?

~5-6ms on A100, not that noticeable

File: onnxruntime/contrib_ops/cuda/transformers/generation_device_helper.cc (hunk 2 above)
Comment (wangyems) [reply onnxruntime/contrib_ops/cuda/transformers/generation_device_helper.cc:86]: Another viable approach is to use "ping-pang" buffer and hide memcpy time. However it only works(show gain) for small inputs thus not included in this PR.

File: onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu [hunk 3]
```diff
@@ -1315,6 +1315,81 @@
...
+                                     int batch_size,
+                                     int num_heads,
+                                     int max_length,
+                                     int head_size,
+                                     int chunk_size,
+                                     cudaStream_t stream) {
+  const int chunked_head_size = head_size / chunk_size;
+  const dim3 block(16, chunked_head_size, chunk_size / 4);
+  const dim3 grid(batch_size, num_heads, (max_length + block.x - 1) / block.x);
+  if (chunk_size == 4) {
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu:1372]: > if (chunk_size == 4) { [](http://example.com/codeflow?start=1&length=23)

there is no need to differentiate half or float. The last dimension keeps 16bytes, which is a float4. So you can convert a half tensor of [B, N, S, H] to float4 tensor with shape [B, N, S, H/8] and a float tensor to float4 tensor with shape [B, N, S, H/4]

File: onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu [hunk 4]
```diff
@@ -1315,6 +1315,81 @@
...
+
+    __syncthreads();
+
+    const int out_offset = b * num_heads * max_length * chunked_head_size * equv_chunk_size +
+                           n * max_length * chunked_head_size * equv_chunk_size +
+                           h2 * max_length * equv_chunk_size +
+                           s * equv_chunk_size +
+                           c;
+
+    out_buffer[out_offset] = tile[tile_x][tile_y];
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu:1357]: neither input nor output are accessed continouesly. Is it efficient?

File: onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu [hunk 5]
```diff
@@ -1315,6 +1315,81 @@
...
+void ReorderPastStatesKernelLauncher(void* out_buffer,
+                                     const void* in_buffer,
+                                     int batch_size,
+                                     int num_heads,
+                                     int max_length,
+                                     int head_size,
+                                     int chunk_size,
+                                     cudaStream_t stream) {
+  const int chunked_head_size = head_size / chunk_size;
+  const dim3 block(16, chunked_head_size, chunk_size / 4);
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu:1370]: >m3 block(16, chunked_head_size, chunk_size / 4); [](http://example.com/codeflow?start=10&length=48)

Maybe consider order of (chunk_size /4, chunked_head_size, 16)

File: onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu (hunk 4 above)
Comment (wangyems) [reply onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu]: after fixing the index, the transpose kernel is even faster than memcpy!
image

**Review by yufenglee — APPROVED** [review]
This is nice! As we synced offline, there will be a following PR to remove most of the memcpy.

**Pull request merged**